libraries/chain/db_market.cpp

```diff
@@ -1192,34 +1217,57 @@ asset database::calculate_market_fee( const asset_object& trade_asset, const ass
    return percent_fee;
 }
 
-asset database::pay_market_fees(const account_object* seller, const asset_object& recv_asset, const asset& receives )
+/***
+ * @brief calculate the margin fee
+ * 
+ * Note that the trade_asset controls the fee percentage, but trade_amount controls the asset the fee is paid in
+ * @param trade_asset the asset that controls the fee
+ * @param trade_amount the asset that the fee should be paid in
+ * @returns the amount to be paid
+ */
+asset database::calculate_margin_fee( const asset_object& trade_asset, const asset& trade_amount)
+{
+   assert( trade_asset.id != trade_amount.asset_id );
+
+   if( !trade_asset.charges_market_fees() )
+      return asset(0, trade_amount.asset_id);
+   if( !trade_asset.options.extensions.value.margin_call_fee_ratio.valid() 
+         || *trade_asset.options.extensions.value.margin_call_fee_ratio == 0 )
+      return asset(0, trade_amount.asset_id);
+
+   auto value = detail::calculate_percent(trade_amount.amount, *trade_asset.options.extensions.value.margin_call_fee_ratio);
```

Comment:
This seems different from the specification.

Reply:
The problem I have is clarifying the specifications. I apologize for the lengthy post, but it is how my mind works. The BSIP says:

> When a margin call trading happens, the buyer sells smartcoin with quantity X and get collaterals in quantity `X*(MSSR-MCFR)/settlement price`, the margin call order owner sells collaterals in quantity `X*MSSR/settlement price` and get smartcoin in quantity X, the delta in paid and received collaterals in quantity `X*MCFR/settlement price` will be paid to the owner of the smartcoin as margin call fee.

My interpretation (which I will attempt to keep updated within this comment, and probably move to my blog to help me in the future):

Doug the Debtor put up collateral of token C which created token D. Doug was charged an operation fee I think, although I am not sure but it is unimportant. We will assume Doug then sold or transferred D ( also unimportant in this context).

Larry the creator of a Limit order would like to sell the desired quantity of token D and receive the desired quantity of token C. Upon placing the order, Larry was charged a limit order create fee, which is not important in this context.
 
When the debt position was created, a call order was created for Doug the debtor. Within that call order is the current debt and collateral balances.

When the settlement price of the price feed reaches a certain level, Doug's position will get called. Such a margin call will generate revenue for the asset issuer, if the issuer has set the new `margin_call_fee_ratio` within that asset's options.

This fee does not affect Larry. He simply placed a limit order on the book of the exchange.

Doug will see the fee if he compares the amount of collateral he received from the transaction with the amount of debt that was paid off by the margin call.

Now the more technical details: 

When will these orders be matched? When the settlement price from the price feed pushes the value of Doug's collateral below the allowed threshold for the debt asset (see `db_market.cpp#database::check_call_orders`).

What will Larry receive? Exactly what he wished in his limit order.

What will Doug receive? Relief from (some or all of) his debt, and some of his collateral back. He will not receive the full value of his collateral, based on the feed's settlement price. Some of the value of the transaction was paid to the issuer of the debt asset as a margin call fee.

Implementation details:

database::check_call_orders (db_market.cpp) knows when to trigger a margin call. When triggered, `fill_call_order` is called, and then `fill_limit_order` is called.

Within `fill_call_order` the fee is taken from the debtor's debt position and given to the asset issuer as a vesting balance. The remaining value of the debt is used to calculate the collateral that the debtor receives. That collateral is then placed in the debtor's available account balance.

Reply:
In looking at the spec, I believe that I need to adjust the "trigger" of the margin call. The spec reads:

> `Margin call order price = settlement price/(MSSR-MCFR)`

Reply:
I may have written something wrong in previous comments. Let's forget them and start here.

With bsip74,
* condition when a call order is possible to be matched does not change, it's still call_price >= feed_price;
* condition where (at what price, aka match_price) the call order be matched does change,
  * previously, if a limit order buys at `feed_price / mssr`, it will match the call order,
  * now, if a limit order buys at `feed_price / (mssr - mcfr)`, it will match the call order,
* when matched, the limit order gets collateral (be filled) at `feed_price / (mssr - mcfr)`, the call order pays (be filled) at `feed_price / mssr`, the difference goes to the debt asset owner;
* the call order can be a maker or a taker, there is a small difference in match_price, iirc maker and taker are handled in different functions.

Reply:
One important thing, the fee is some amount of the collateral asset, so it can not go to the `accumulated_fees` field of the debt asset, which means we likely can not reuse the code of `pay_market_fees`. It can not go to the `accumulated_fees` field of the collateral asset either, since the asset can be owned by someone else.

In addition, a complex scenario is the debt asset can change from one collateral to another, so we need to store the fees in a multi-asset container. Adding new fields into asset_object or another object means we need to add new operations to claim them, that's too much work to implement.

Reply:
Another thing in mind: `mssr-mcfr` in the specification, and we don't allow something for nonthing, so it's implied that `mssr>mcfr`. However, mssr is updated by price feed producers and an internal timer, mcfr is updated by asset owner, so likely we can't validate when one of them is updated. So we need to cap the value of mcfr when use it in calculation.

Reply:
@abitmore 

I believe I am clear that this change modifies which limit orders qualify to be matched. 

To be clear, let me walk through what I am seeing in db_market.cpp::check_call_orders. And just focusing on the price the limit order will transact at...

Prior to this change, the limit order was filled at the price dictated by the order itself:

`price match_price  = limit_order.sell_price;`

Based on what I'm reading in your comment above, after the hardfork, the limit order will transact at a price based on feed price, MSSR and MCFR. That will result in the limit order receiving a price potentially very different than their order.

Or it could be that I am misreading your comment above, and you were only presenting the extreme example of when a limit order just happens to match right at the new calculated price for what qualifies a match, which would mean the order gets exactly what it asked for.

Or I am making assumptions in the code that are incorrect. 

Would you clarify please?

Reply:
Please keep in mind that match price is always price of the maker, see [BSIP 32](https://github.com/bitshares/bsips/blob/master/bsip-0032.md).
* In check_call_orders, the call order is the taker and the limit order is the maker, so the match price will always be the price of the limit order.
* In another function where the call order is the maker and the limit order is the taker, the match price will be settlement_price / (mssr-mcfr).

Reply:
Thank you for the reference to BSIP 32. There are some inconsistencies in that BSIP that make it difficult to understand, but I believe the gist is as you say, when the limit order is the maker, the maker price is what is used.

So that leads me to believe that BSIP 74 applies only when the call order is the maker. Hence the margin call fee will not apply when the limit order is the maker. Is that correct?

Reply:
IMO the margin call fee should always apply, although how much to pay is not clearly described in the BSIP when Peter approved the pull request.

In case when the call order is taker, I think it's fine to charge the amount `filled_debt * mcfr / feed_price` from the call order as fee.